### PR TITLE
OEM-IBM: PDR: Add Real SAI sensor and effecter Support

### PR DESCRIPTION
**Pull request by SrideviRamesh**
Signed-off-by: Sridevi Ramesh <sridevra@in.ibm.com>

Real SAI effecter
================

{
    "nextRecordHandle": 138,
    "responseCount": 30,
    "recordHandle": 137,
    "PDRHeaderVersion": 1,
    "PDRType": "State Effecter PDR",
    "recordChangeNumber": 0,
    "dataLength": 16,
    "PLDMTerminusHandle": 1,
    "effecterID": 24,
    "entityType": "[Physical] OEM IBM Real SAI(OEM)",
    "entityInstanceNumber": 1,
    "containerID": 1,
    "effecterSemanticID": 0,
    "effecterInit": "noInit",
    "effecterDescriptionPDR": false,
    "compositeEffecterCount": 1,
    "stateSetID[0]": "Operational Fault Status(10)",
    "possibleStatesSize[0]": 1,
    "possibleStates[0]": [
        "Normal(1)"
    ]
}

Real SAI sensor
===============

{
    "nextRecordHandle": 139,
    "responseCount": 28,
    "recordHandle": 138,
    "PDRHeaderVersion": 1,
    "PDRType": "State Sensor PDR",
    "recordChangeNumber": 0,
    "dataLength": 14,
    "PLDMTerminusHandle": 1,
    "sensorID": 22,
    "entityType": "[Physical] OEM IBM Real SAI(OEM)",
    "entityInstanceNumber": 1,
    "containerID": 1,
    "sensorInit": "noInit",
    "sensorAuxiliaryNamesPDR": false,
    "compositeSensorCount": 1,
    "stateSetID[0]": "Operational Fault Status(10)",
    "possibleStatesSize[0]": 2,
    "possibleStates[0]": [
        "Normal(1)",
        "Stressed(2)"
    ]
}

Entity association PDR
======================

{
    "nextRecordHandle": 77,
    "responseCount": 56,
    "recordHandle": 76,
    "PDRHeaderVersion": 1,
    "PDRType": "Entity Association PDR",
    "recordChangeNumber": 0,
    "dataLength": 46,
    "containerID": 1,
    "associationType": "Physical",
    "containerEntityType": "[Physical] System (Logical)",
    "containerEntityInstanceNumber": 1,
    "containerEntityContainerID": 0,
    "containedEntityCount": 6,
    "containedEntityType[1]": "[Physical] System chassis (main enclosure)",
    "containedEntityInstanceNumber[1]": 1,
    "containedEntityContainerID[1]": 1,
    "containedEntityType[2]": "[Physical] OEM IBM Firmware Update(OEM)",
    "containedEntityInstanceNumber[2]": 1,
    "containedEntityContainerID[2]": 1,
    "containedEntityType[3]": "[Physical] OEM IBM Real SAI(OEM)",
    "containedEntityInstanceNumber[3]": 1,
    "containedEntityContainerID[3]": 1,
    "containedEntityType[4]": "[Physical] Battery",
    "containedEntityInstanceNumber[4]": 1,
    "containedEntityContainerID[4]": 1,
    "containedEntityType[5]": "[Physical] Indicator",
    "containedEntityInstanceNumber[5]": 1,
    "containedEntityContainerID[5]": 1,
    "containedEntityType[6]": "[Physical] Indicator",
    "containedEntityInstanceNumber[6]": 2,
    "containedEntityContainerID[6]": 1
}

Verify Real Sensor and effecter
==============================

 /tmp/pldmtool platform GetStateSensorReadings -i 22 -r 0
{
    "compositeSensorCount": 1,
    "sensorOpState[0]": "Sensor Enabled",
    "presentState[0]": "Sensor Normal",
    "previousState[0]": "Sensor Unknown",
    "eventState[0]": "Sensor Normal"
}
 busctl set-property xyz.openbmc_project.LED.GroupManager /xyz/openbmc_project/led/groups/platform_system_attention_indicator xyz.openbmc_project.Led.Group Asserted b true

/tmp/pldmtool platform GetStateSensorReadings -i 22 -r 0                                                            {
    "compositeSensorCount": 1,
    "sensorOpState[0]": "Sensor Enabled",
    "presentState[0]": "Sensor Normal",
    "previousState[0]": "Sensor Unknown",
    "eventState[0]": "Sensor Warning"
}

 /tmp/pldmtool platform SetStateEffecterStates -i 24 -c 1 -d 1 1
{
    "Response": "SUCCESS"
}
/tmp/pldmtool platform GetStateSensorReadings -i 22 -r 0
{
    "compositeSensorCount": 1,
    "sensorOpState[0]": "Sensor Enabled",
    "presentState[0]": "Sensor Normal",
    "previousState[0]": "Sensor Unknown",
    "eventState[0]": "Sensor Normal"

### CODE REVIEW

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp [hunk 1]
```diff
@@ -182,6 +193,7 @@ int pldm::responder::oem_ibm_platform::Handler::
             }
             else if (stateSetId == PLDM_OEM_IBM_SLOT_ENABLE_EFFECTER_STATE)
             {
+                std::cerr << "Effecter id : " << effecterId << std::endl;
```
Comment (manojkiraneda) [inline oem/ibm/libpldmresponder/oem_ibm_handler.cpp:196]: please remove this trace , as this is not related

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp [hunk 2]
```diff
@@ -1227,7 +1346,8 @@ void pldm::responder::oem_ibm_platform::Handler::handleBootTypesAtPowerOn()
         }
         catch (const std::exception& e)
         {
-            return;
+            std::cerr << "Failed to fetch the restartCause, ERROR=" << e.what()
+                      << "\n";
```
Comment (manojkiraneda) [inline oem/ibm/libpldmresponder/oem_ibm_handler.cpp:1350]: this is an unrelated change , please remove

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp [hunk 3]
```diff
@@ -1301,6 +1421,56 @@ void pldm::responder::oem_ibm_platform::Handler::handleBootTypesAtChassisOff()
...
+        pldm::utils::DBusHandler().setDbusProperty(dbuspartitionMapping, false);
+
+        pldm::utils::DBusMapping dbusplatformMapping{
+            "/xyz/openbmc_project/led/groups/platform_system_attention_indicator",
+            "xyz.openbmc_project.Led.Group", "Asserted", "bool"};
+        pldm::utils::DBusHandler().setDbusProperty(dbusplatformMapping, false);
+    }
+    catch (const std::exception& e)
+    {
+        std::cerr << "Failed to set the Real SAI effecter" << std::endl;
```
Comment (manojkiraneda) [inline oem/ibm/libpldmresponder/oem_ibm_handler.cpp:1440]: any of the above dbus call fails , we enter the exception.

For whatever reason, if one fails, how will you know which one failed? , so please print the patch of dbus object and also print which exception you got and what is the reason.

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp [hunk 4]
```diff
@@ -1301,6 +1421,56 @@ void pldm::responder::oem_ibm_platform::Handler::handleBootTypesAtChassisOff()
...
+            platformSAIObjectPath, saiPropertyName, saiInterface);
+
+        if (isPartitionSAIOn || isPlatformSAIOn)
+        {
+            return PLDM_SENSOR_WARNING;
+        }
+    }
+    catch (const std::exception& e)
+    {
+        std::cerr << "Failed to fetch Real SAI sensor status" << std::endl;
```
Comment (manojkiraneda) [inline oem/ibm/libpldmresponder/oem_ibm_handler.cpp:1469]: same comments as above.

Comment (manojkiraneda) [reply oem/ibm/libpldmresponder/oem_ibm_handler.cpp:1469]: and also what will you return when you get an exception?

File: pldmtool/oem/ibm/oem_ibm_state_set.hpp [hunk 5]
```diff
@@ -55,6 +55,8 @@ extern const std::map<uint8_t, std::string> OemIBMEntityType{
      "OEM IBM Firmware Update"},
     {PLDM_OEM_ENTITY_TYPE_START, "OEM IBM Entity Type Start"},
     {PLDM_OEM_ENTITY_TYPE_END, "OEM IBM Entity Type End"},
+    {pldm::responder::oem_ibm_platform::PLDM_OEM_IBM_ENTITY_REAL_SAI,
```
Comment (manojkiraneda) [inline pldmtool/oem/ibm/oem_ibm_state_set.hpp:58]: `pldm::responder::oem_ibm_platform::` 
is the name space really required ?

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp (hunk 1 above)
Comment (SrideviRamesh) [reply oem/ibm/libpldmresponder/oem_ibm_handler.cpp:196]: done

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp (hunk 2 above)
Comment (SrideviRamesh) [reply oem/ibm/libpldmresponder/oem_ibm_handler.cpp:1350]: done

File: pldmtool/oem/ibm/oem_ibm_state_set.hpp (hunk 5 above)
Comment (SrideviRamesh) [reply pldmtool/oem/ibm/oem_ibm_state_set.hpp:58]: If I dont mention I get this  error: ‘PLDM_OEM_IBM_ENTITY_REAL_SAI’ was not declared in this scope

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp (hunk 4 above)
Comment (SrideviRamesh) [reply oem/ibm/libpldmresponder/oem_ibm_handler.cpp:1469]: I am throwing error msg.. I was initially returning none... But i got the comment to remove as not required.. btw also checked in the same file mostly nothing is returned on a error statement

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp (hunk 3 above)
Comment (SrideviRamesh) [reply oem/ibm/libpldmresponder/oem_ibm_handler.cpp:1440]: As discussed added separate try block and changed the error statement accordingly

**Review by jaypadath — APPROVED** [review]

**Review by manojkiraneda — APPROVED** [review]

**Review by sampmisr — APPROVED** [review]
